transports/quic/src/certificate.rs

```diff
@@ -0,0 +1,235 @@
+//!
+//! This crate uses the `log` crate to emit log output.  Events that will occur normally are output
+//! at `trace` level, while “expected” error conditions (ones that can result during correct use of the
+//! library) are logged at `debug` level.
+use log::{trace, warn};
+const LIBP2P_OID: &[u64] = &[1, 3, 6, 1, 4, 1, 53594, 1, 1];
+const LIBP2P_SIGNING_PREFIX: [u8; 21] = *b"libp2p-tls-handshake:";
+pub const LIBP2P_SIGNING_PREFIX_LENGTH: usize = LIBP2P_SIGNING_PREFIX.len();
+const LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH: usize = 65;
+static LIBP2P_SIGNATURE_ALGORITHM: &rcgen::SignatureAlgorithm = &rcgen::PKCS_ECDSA_P256_SHA256;
+use libp2p_core::identity;
+
+fn encode_signed_key(public_key: identity::PublicKey, signature: &[u8]) -> rcgen::CustomExtension {
+    let public_key = public_key.into_protobuf_encoding();
+    let contents = yasna::construct_der(|writer| {
+        writer.write_sequence(|writer| {
+            writer
+                .next()
+                .write_bitvec_bytes(&public_key, public_key.len() * 8);
+            writer
+                .next()
+                .write_bitvec_bytes(signature, signature.len() * 8);
+        })
+    });
+    rcgen::CustomExtension::from_oid_content(LIBP2P_OID, contents)
+}
+
+fn gen_signed_keypair(keypair: &identity::Keypair) -> (rcgen::KeyPair, rcgen::CustomExtension) {
+    let temp_keypair = rcgen::KeyPair::generate(&LIBP2P_SIGNATURE_ALGORITHM)
+        .expect("we pass valid parameters, and assume we have enough memory and randomness; qed");
+    let mut signing_buf =
+        [0u8; LIBP2P_SIGNING_PREFIX_LENGTH + LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH];
+    let public = temp_keypair.public_key_raw();
+    assert_eq!(
+        public.len(),
+        LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH,
+        "ECDSA public keys are 65 bytes"
+    );
+    signing_buf[..LIBP2P_SIGNING_PREFIX_LENGTH].copy_from_slice(&LIBP2P_SIGNING_PREFIX[..]);
+    signing_buf[LIBP2P_SIGNING_PREFIX_LENGTH..].copy_from_slice(public);
+    let signature = keypair.sign(&signing_buf).expect("signing failed");
+    (
+        temp_keypair,
+        encode_signed_key(keypair.public(), &signature),
+    )
+}
+
+pub fn make_cert(keypair: &identity::Keypair) -> rcgen::Certificate {
+    let mut params = rcgen::CertificateParams::new(vec![]);
+    let (cert_keypair, libp2p_extension) = gen_signed_keypair(keypair);
+    params.custom_extensions.push(libp2p_extension);
+    params.alg = &LIBP2P_SIGNATURE_ALGORITHM;
+    params.key_pair = Some(cert_keypair);
+    rcgen::Certificate::from_params(params)
+        .expect("certificate generation with valid params will succeed; qed")
+}
+
+/// Read a bitvec into a vector of bytes.  Requires the bitvec to be a whole number of bytes.
+fn read_bitvec(reader: &mut yasna::BERReaderSeq) -> Result<Vec<u8>, yasna::ASN1Error> {
+    let (value, bits) = reader.next().read_bitvec_bytes()?;
+    // be extra careful regarding overflow
+    if bits.trailing_zeros() >= 3 {
```

Comment:
What is this for?
If it is checking that `bits` is a multiple of 8, I'd do `(bits % 8) == 0` instead and let the compiler optimize things out.